File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Services/PanoClientService+App.m
//
//  PanoClientService+App.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/15.
//

#import "PanoClientService+App.h"
#import "AppConfig.h"
#import "PanoClientService+OpenApi.h"

static NSString * kUserNameKey = @"com.pano.UserName";
static NSString * kRoomIdKey = @"com.pano.RoomId";
static NSString * kUserIdKey = @"com.pano.UserId";
static NSString * kAudioDebugKey = @"com.pano.AudioDebug";

static NSString * kAudioAdvanceProcess = @"com.pano.audioAdvanceProcess";
static NSString * kAudioVoiceChanger = @"com.pac.audioVoiceChanger";

static NSString * kHttpHeader = @"http";
static NSString * kHttpsHeader = @"https";
static NSString * kUrlCheckUpdateApi = @"app/checkUpdate";
static NSString * kUrlCheckLocationApi = @"app/checkLoc";
static NSString * kUrlLoginApi = @"app/token/v2";


@implementation PanoClientService (App)

- (void)savePreferences {
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    [userDefaults setObject:self.userName forKey:kUserNameKey];
    [userDefaults setObject:self.roomId forKey:kRoomIdKey];
    [userDefaults setInteger:self.userId forKey:kUserIdKey];
    [userDefaults setBool:self.audioDebug forKey:kAudioDebugKey];
    
    [userDefaults setInteger:self.audioVoiceChanger forKey:kAudioVoiceChanger];
    [userDefaults setBool:self.audioAdvanceProcess forKey:kAudioAdvanceProcess];
    
    if (self.audioDebug) {
        [self.audioService startAudioDump:-1];
    } else {
        [self.audioService stopAudioDump];
    }
}

- (void)loadPreferences {
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    self.userName = [userDefaults stringForKey:kUserNameKey];
    self.roomId = [userDefaults stringForKey:kRoomIdKey];
    self.userId = [userDefaults integerForKey:kUserIdKey];
    self.audioDebug = [userDefaults boolForKey:kAudioDebugKey];
    if ([userDefaults valueForKey:kAudioAdvanceProcess]) {
        self.audioAdvanceProcess = [userDefaults boolForKey:kAudioAdvanceProcess];
    }
    if ([userDefaults valueForKey:kAudioVoiceChanger]) {
        self.audioVoiceChanger = [userDefaults integerForKey:kAudioVoiceChanger];
    }
}

- (void)uploadlogsAutomatically {
    if (self.uploadlogType != PanoMsgNone) {
        self.uploadlogType = PanoMsgNone;
        NSString *detail = [NSString stringWithFormat:@"%@ 自动上传日志: %@", self.config.userName ,self.uploadlogMessage];
        [self sendFeedback:kPanoFeedbackAudio detail:detail uploadLogs:true contact:@""];
    }
}
- (void)sendFeedback:(PanoFeedbackType)type
              detail:(NSString *)detail
          uploadLogs:(BOOL)uploadLogs
             contact:(NSString *)contact {
    PanoFeedbackInfo * info = [PanoFeedbackInfo new];
    info.type = type;
    info.productName = [[self class] productName];
    info.detailDescription = detail;
    info.contact = contact;
    info.extraInfo = @"";
    info.uploadLogs = uploadLogs;
    [self.engineKit sendFeedback:info];
}

- (void)notifyOthersUploadLogs:(PanoFeedbackType)type
                       message:(NSString *)message
{
    self.uploadlogMessage = message;
    [self broadcastMessage:[[[PanoCmdMessage alloc] initWithCmd:uploadAudioLog data:nil] toDictionary]];
}

+ (NSString *)productName {
    return [NSBundle.mainBundle.infoDictionary objectForKey:@"CFBundleName"];
}

+ (NSString *)productVersion {
    NSString * appVersion = [[self class] productShortVersion];
    NSString * sdkVersion = [PanoRtcEngineKit getSdkVersion];
    return [NSString stringWithFormat:@"v%@ (%@)", appVersion, sdkVersion];
}

+ (NSString *)productShortVersion {
    return [NSBundle.mainBundle.infoDictionary objectForKey:@"CFBundleShortVersionString"];
}

- (void)fetchTokenWithCompletionHandler:(void (^)(NSString *token, UInt64 hostUserId, NSError * error))completionHandler {
    /**
     *  SDK 到 Pano Cloud 的所有交互都需要使用 Token，Token的生成方式有两种。
     *   第一种是 App Server 通过 RESTful API 向 Pano 服务器获取。
     *   第二种是 App Server 按照 Pano 定义的规则本地生成
     *   请参考 https://developer.pano.video/restful/authtoken/
     *
     *   第三种 调试阶段 可以临时 Token, 可以在控制台创建应用
     *   请参考 https://console.pano.video/#/user/login
     *
     *   加入房间后需要获取到 "Token" 和  "主播的 UserId"，需要 客户的App Server 提供接口
     *
     *   创建语聊房  调试可以返回 "临时Token"  和 "当前用户的Userid"
     *   加入语聊房  调试可以返回 "临时Token"  和 "主播UserId"
     *
     *   客户的App Server 需要保证
     *   1. 同一个房间ID，同一时间 只能有一个主播加入
     */
    completionHandler(<#T##Token: String##String#>, <#T##主播的UserId: String##String#>, nil);
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Service/PanoClientService+App.swift
//
//  PanoClientService+App.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2021/6/11.
//

import Foundation

extension PanoClientService {
    
    static let kUserUniqueIDKey = "UUID";
    static let kUserNameKey = "com.pac.UserName";
    static let kDebugKey = "com.pac.debug"
    static let kRoomIdKey = "com.pac.roomID"
    static let kUserIdKey = "com.pac.userID"
    
    static let kAudioAdvanceProcess = "com.pac.audioAdvanceProcess"
    static let kAudioVoiceChanger = "com.pac.audioVoiceChanger"
    
    @objc static var uuid: String = ""
    @objc static var userName = ""
    @objc static var debug: Bool = false
    @objc static var roomId = ""
    @objc static var userId: UInt64 = 0
    @objc static var audioAdvanceProcess: Bool = true
    static var audioVoiceChanger: PanoAudioVoiceChangerOption = .voiceChangerNone
    
    @objc func sendFeedback(type: PanoFeedbackType,
                            detail: String? ,
                            uploadLogs: Bool,
                            contact: String?) {
        let info = PanoFeedbackInfo()
        info.type = type
        info.productName = AppInfo.productName
        info.detailDescription = detail ?? ""
        info.contact = contact
        info.extraInfo = PanoClientService.uuid
        info.uploadLogs = uploadLogs
        self.engineKit?.sendFeedback(info)
    }
    
    @objc func productVersion() -> String {
        let appVersion = AppInfo.appVersion;
        let sdkVersion = PanoRtcEngineKit.getSdkVersion();
        return "v\(appVersion) \(sdkVersion)"
    }
    
    open class func checkUUID() {
        let uuId = UserDefaults.standard.string(forKey: PanoClientService.kUserUniqueIDKey)
        if let id = uuId {
            PanoClientService.uuid = id
        } else {
            PanoClientService.uuid = NSUUID().uuidString
            UserDefaults.standard.setValue(PanoClientService.uuid, forKey: PanoClientService.kUserUniqueIDKey)
        }
    }
    
    open class func loadPreferences() {
        checkUUID()
        PanoClientService.userName = UserDefaults.standard.string(forKey: PanoClientService.kUserNameKey) ?? ""
        PanoClientService.debug = UserDefaults.standard.bool(forKey: PanoClientService.kDebugKey)
        PanoClientService.roomId = UserDefaults.standard.string(forKey: PanoClientService.kRoomIdKey) ?? ""
        PanoClientService.userId = UInt64(UserDefaults.standard.integer(forKey: PanoClientService.kUserIdKey))
        
        if let _ = UserDefaults.standard.value(forKey: PanoClientService.kAudioAdvanceProcess) {
            audioAdvanceProcess = UserDefaults.standard.bool(forKey: PanoClientService.kAudioAdvanceProcess)
        }
        
        PanoClientService.audioVoiceChanger = PanoAudioVoiceChangerOption(rawValue: UserDefaults.standard.integer(forKey: PanoClientService.kAudioVoiceChanger)) ?? .voiceChangerNone
    }
    
    @objc open class func savePreferences() {
        UserDefaults.standard.setValue(PanoClientService.userName, forKey: PanoClientService.kUserNameKey)
        UserDefaults.standard.setValue(PanoClientService.debug, forKey: PanoClientService.kDebugKey)
        UserDefaults.standard.setValue(PanoClientService.roomId, forKey: PanoClientService.kRoomIdKey)
        UserDefaults.standard.setValue(PanoClientService.userId, forKey: PanoClientService.kUserIdKey)
        
        UserDefaults.standard.setValue(PanoClientService.audioAdvanceProcess, forKey: PanoClientService.kAudioAdvanceProcess)
        UserDefaults.standard.setValue(PanoClientService.audioVoiceChanger.rawValue, forKey: PanoClientService.kAudioVoiceChanger)
        
        if PanoClientService.debug {
            PanoAudioService.service()?.startAudioDump()
        } else {
            PanoAudioService.service()?.stopAudioDump()
        }
    }
    
//    自动上传日志
    func uploadlogsAutomatically() {
        if uploadlogType != nil {
            uploadlogType = nil;
            sendFeedback(type: PanoFeedbackType.audio, detail: "\(config.userName) 自动上传日志: " + (uploadlogMessage ?? ""), uploadLogs: true, contact: nil)
        }
    }
    
    @objc func notifyOthersUploadLogs(type: PanoFeedbackType, message: String) {
        uploadlogMessage = message
        _ = self.broadcastMessage(msg: PanoCmdMessage(cmd: .uploadAudioLog).toDictionary())
    }
    
    @objc open class func setAudioVoiceChangerMode(mode: PanoAudioVoiceChangerOption) {
        PanoClientService.audioVoiceChanger = mode
        PanoClientService.rtcEngineKit()?.setOption(mode.rawValue as NSObject, for: .optionAudioVoiceChangerMode)
    }
    
    public static func checkAppVersion(handler: @escaping (_ forceUpdate: Bool) -> Void) {
        // TODO 
    }
}
